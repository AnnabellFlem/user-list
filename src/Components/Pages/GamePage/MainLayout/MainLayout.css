@import '../../../../Styles/colors.css';
@import '../../../../Styles/sizes.css';

.MainLayout {
  grid-area: main;
  height: 100%;
  background-color: var(--black-5);
}

.MainLayout__answers-fieldset {
  margin: 0;
  padding: 0;
  border: 0;
  height: 100%;
  flex-wrap: wrap;
  display: flex;
  flex-direction: column;
}

.MainLayout__answers-form {
  display: flex;
  flex-direction: column;
  padding-left: 80px;
  padding-bottom: 122px;
  padding-top: 133px;
  height: 100%;
  background-color: transparent;
}

.MainLayout__answers-text {
  max-width: 624px;
  font-weight: 600;
  font-size: var(--font-m);
  line-height: var(--lh-l);
  padding: 0;
}

.MainLayout__answers-list {
  display: flex;
  flex-wrap: wrap;
  margin: auto 0 0;
  max-width: 870px;
}

.MainLayout__answers-item {
  width: 420px;
}

.MainLayout__answers-item:nth-child(n + 3) {
  margin-top: 32px;
}

.MainLayout .btn {
  margin-top: 32px;
  width: max-content;
}

.MainLayout--loading {
  pointer-events: none;
  user-select: none;
}

@media (max-width: 1400px) {
  .MainLayout__answers-item {
    width: auto;
  }
}

@media (max-width: 1350px) {
  .MainLayout__answers-item {
    width: 420px;
  }

  .MainLayout__answers-item:nth-child(n + 1) {
    margin-top: 32px;
  }
}

@media (max-width: 920px) {
  .MainLayout__answers-form {
    padding: 136px 0 40px 0;
  }

  .MainLayout__answers-fieldset {
    z-index: 2;
    display: flex;
    flex-direction: column;
  }

  .MainLayout__answers-text {
    text-align: center;
    padding: 0 16px;
    align-self: center;
  }

  .MainLayout__answers-item {
    flex-basis: 100%;
    width: 100%;
  }

  .MainLayout__answers-item + .MainLayout__answers-item {
    margin-top: 8px;
  }

  .MainLayout .btn {
    z-index: 2;
    align-self: center;
  }
}

@media (max-width: 600px) {
  .MainLayout .btn {
    width: calc(100% - 32px);
  }

  .MainLayout__answers-text {
    font-size: var(--font-s);
    line-height: var(--lh-s-plus);
  }
}

@media (max-width: 350px) {
  .MainLayout__answers-form {
    padding-top: 102px;
  }
}
