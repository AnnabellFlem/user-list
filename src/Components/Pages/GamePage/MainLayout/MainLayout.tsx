import React, { useEffect, useState } from 'react'
import './MainLayout.css'
import AnswerCell from '../../../BaseComponents/AnswerCell'
import Button from '../../../BaseComponents/Button'
import { questionArray, moneyArray } from '../../../../Services/mockData'
import useMedia from 'use-media'
import { MEDIA_QUERY_MOBILE } from '../../../../Constants/MediaQueries'
import { AnswersType } from '../../../../Types'
import classNames from 'classnames'

type AnswersArrayType = Array<AnswersType>

type MainLayoutType = {
  handleCurrentQuestion: (question: number) => void
  currentNumberQuestion: number
  handleEndGame: (isEndGame: boolean) => void
}

const MainLayout: React.FC<MainLayoutType> = ({
  handleCurrentQuestion,
  currentNumberQuestion,
  handleEndGame,
}) => {
  const isWide = useMedia(MEDIA_QUERY_MOBILE)
  const lastRound = moneyArray.length
  const [answersArray, setAnswersArray] = useState<Array<string>>([])
  const [resultArray, setResultArray] = useState<AnswersArrayType>([])
  const [rightAnswer, setRightAnswer] = useState<boolean | null>()

  const currentQuestion = questionArray[currentNumberQuestion]

  const defaultMainLayoutClassName = 'MainLayout'
  const mainLayoutClassNames = classNames(defaultMainLayoutClassName, {
    [`${defaultMainLayoutClassName}--loading`]: rightAnswer !== null,
  })

  const arrayEquals = (arrayOne: Array<any>, arrayTwo: Array<any>) => {
    return (
      Array.isArray(arrayOne) &&
      Array.isArray(arrayTwo) &&
      arrayOne.length === arrayTwo.length &&
      arrayOne.every((val, index) => val === arrayTwo[index])
    )
  }

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const currID = e.currentTarget.id
    const index = answersArray.indexOf(currID)
    if (index > -1) {
      answersArray.splice(index, 1)
      return setAnswersArray(answersArray)
    }
    return setAnswersArray([...answersArray, currID])
  }

  const handleCheck = () => {
    const rightAns = currentQuestion.answers.filter(item => item.right)

    const userAns = currentQuestion.answers.filter(item => {
      if (answersArray.includes(item.id)) {
        return item
      }
      return null
    })

    setResultArray([...rightAns, ...userAns.filter(item => !item.right)])

    if (arrayEquals(rightAns, userAns)) {
      setRightAnswer(true)
    } else return setRightAnswer(false)
  }

  useEffect(() => {
    if (rightAnswer !== null) {
      const timeout = setTimeout(() => {
        if (rightAnswer) {
          setAnswersArray([])
          setResultArray([])
          currentNumberQuestion === lastRound - 1
            ? handleEndGame(true)
            : handleCurrentQuestion(currentNumberQuestion + 1)
        } else handleEndGame(false)
      }, 2000)
      return () => clearTimeout(timeout)
    }
    // eslint-disable-next-line
  }, [rightAnswer])

  useEffect(() => {
    handleCurrentQuestion(currentNumberQuestion)
    setRightAnswer(null)
  }, [currentNumberQuestion, handleCurrentQuestion])

  const RenderList = (array: AnswersArrayType, resArr: AnswersArrayType) => {
    return (
      <>
        {array.map(item => {
          const currItem = resArr.includes(item)
          return (
            <li
              key={item.id}
              className={`${defaultMainLayoutClassName}__answers-item`}
            >
              {isWide ? (
                <AnswerCell
                  variant={item.variant}
                  id={item.id}
                  name="answer"
                  correct={currItem ? item.right : undefined}
                  wrong={currItem ? !item.right : undefined}
                  onChange={handleChange}
                  checked={answersArray.includes(item.id)}
                >
                  {item.answer}
                </AnswerCell>
              ) : (
                <AnswerCell
                  variant={item.variant}
                  sizeCell="s"
                  id={item.id}
                  correct={currItem ? item.right : undefined}
                  wrong={currItem ? !item.right : undefined}
                  name="answer"
                  onChange={handleChange}
                  checked={answersArray.includes(item.id)}
                >
                  {item.answer}
                </AnswerCell>
              )}
            </li>
          )
        })}
      </>
    )
  }

  return (
    <main>
      <section className={mainLayoutClassNames}>
        <form className={`${defaultMainLayoutClassName}__answers-form`}>
          <div
            role="group"
            className={`${defaultMainLayoutClassName}__answers-fieldset`}
          >
            <legend className={`${defaultMainLayoutClassName}__answers-text`}>
              {currentQuestion.question}
            </legend>
            <ul className={`${defaultMainLayoutClassName}__answers-list`}>
              {RenderList(currentQuestion.answers, resultArray)}
            </ul>
          </div>
          {isWide ? (
            <Button type="button" onClick={handleCheck}>
              Check
            </Button>
          ) : (
            <Button type="button" size="s" onClick={handleCheck}>
              Check
            </Button>
          )}
        </form>
      </section>
    </main>
  )
}

export default MainLayout
