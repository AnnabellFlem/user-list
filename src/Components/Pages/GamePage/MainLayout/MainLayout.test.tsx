import React from 'react'
import { mount } from 'enzyme'
import MainLayout from './MainLayout'

const MOCK_HANDLE_CURR_QUESTION = jest.fn()
const MOCK_HANDLE_END_GAME = jest.fn()

describe('<MainLayout /> test', () => {
  let wrapper: any

  beforeAll(() => {
    wrapper = mount(
      <MainLayout
        currentNumberQuestion={0}
        handleEndGame={MOCK_HANDLE_END_GAME}
        handleCurrentQuestion={MOCK_HANDLE_CURR_QUESTION}
      />,
    )
  })

  it('should to match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Component should render without crashing', () => {
    expect(wrapper.exists()).toBeTruthy()
  })
})
