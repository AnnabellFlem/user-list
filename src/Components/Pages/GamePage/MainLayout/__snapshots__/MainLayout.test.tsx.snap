// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainLayout /> test should to match snapshot 1`] = `ReactWrapper {}`;
